File: src/lib/missions/handlers/ACSDefendMission.ts
/**
 * ACSDefendMission Handler
 *
 * Handles ACS defense missions where fleets are sent to defend an ally's planet.
 * The defending fleet joins the planet's defense during any attack.
 *
 * Key Features:
 * - Fleet stays at the target planet for the hold time
 * - Participates in any battle during that time
 * - Returns home after hold time expires or after battle
 */

import type { MissionType } from '@/types/database'
import { BaseMission } from '../BaseMission'
import {
  MissionContext,
  MissionArrivalResult,
  MissionReturnResult,
  Resources,
  ShipCounts,
  emptyResources,
  emptyShipCounts,
  SHIP_KEYS,
  hasResources,
} from '../types'

export class ACSDefendMission extends BaseMission {
  readonly missionType: MissionType = 'acs_defend'
  readonly hasReturn: boolean = true
  readonly name: string = 'ACS Defend'

  /**
   * Process ACS defend mission arrival
   *
   * When a fleet arrives to defend:
   * 1. Register the fleet as stationed at the target planet
   * 2. The fleet will be included in any battles at that location
   * 3. After hold time (or after battle), fleet returns home
   */
  async processArrival(context: MissionContext): Promise<MissionArrivalResult> {
    const { mission, targetPlanet } = context

    // Target planet must exist
    if (!targetPlanet) {
      const ships = this.getShips(mission)
      const resources = this.getResources(mission)

      const message = this.createMessage(
        mission.user_id,
        'battle',
        'ACS Defense Mission Failed',
        `Your defense fleet arrived at ${this.formatCoords(
          mission.destination_galaxy,
          mission.destination_system,
          mission.destination_position
        )} but found no planet. The fleet is returning.`
      )

      return this.successArrival(true, resources, ships, [message], [])
    }

    // Cannot defend your own planets (use deployment instead)
    if (targetPlanet.user_id === mission.user_id) {
      const ships = this.getShips(mission)
      const resources = this.getResources(mission)

      const message = this.createMessage(
        mission.user_id,
        'battle',
        'ACS Defense Error',
        'You cannot send an ACS defense to your own planet. Use deployment instead.'
      )

      return this.successArrival(true, resources, ships, [message], [])
    }

    // Check if there's an ACS operation this fleet belongs to
    const { data: participant } = await this.supabase
      .from('acs_participants')
      .select('*, acs_operations(*)')
      .eq('fleet_mission_id', mission.id)
      .single()

    const ships = this.getShips(mission)
    const resources = this.getResources(mission)

    if (participant) {
      // Part of an ACS operation
      await this.supabase
        .from('acs_participants')
        .update({
          status: 'arrived',
          arrival_time: new Date().toISOString(),
        })
        .eq('id', participant.id)

      const holdTime = participant.acs_operations?.hold_time || 30

      // Create notification for defender
      const defenderMessage = this.createMessage(
        targetPlanet.user_id,
        'battle',
        'Allied Fleet Arrived',
        `An allied fleet has arrived to defend your planet at ${this.formatCoords(
          mission.destination_galaxy,
          mission.destination_system,
          mission.destination_position
        )}.\n\nFleet will stay for ${holdTime} seconds.`
      )

      // Create notification for the fleet owner
      const ownerMessage = this.createMessage(
        mission.user_id,
        'battle',
        'Defense Fleet Stationed',
        `Your fleet has arrived at ${this.formatCoords(
          mission.destination_galaxy,
          mission.destination_system,
          mission.destination_position
        )} and is now defending the planet.\n\nFleet will return after ${holdTime} seconds.`
      )

      // The fleet doesn't return immediately - it waits
      // The mission processor will handle the return after hold time
      // For now, we mark it as not returning to keep it stationed
      return this.successArrival(false, resources, ships, [defenderMessage, ownerMessage], [])
    }

    // Standalone defense (not part of ACS operation)
    // Fleet stations at the planet temporarily

    const defenderMessage = this.createMessage(
      targetPlanet.user_id,
      'battle',
      'Allied Fleet Arrived',
      `An allied fleet has arrived to defend your planet at ${this.formatCoords(
        mission.destination_galaxy,
        mission.destination_system,
        mission.destination_position
      )}.`
    )

    const ownerMessage = this.createMessage(
      mission.user_id,
      'battle',
      'Defense Fleet Stationed',
      `Your fleet has arrived at ${this.formatCoords(
        mission.destination_galaxy,
        mission.destination_system,
        mission.destination_position
      )} and is now defending the planet.`
    )

    // After arrival, fleet will eventually return
    // This is handled by hold time logic
    return this.successArrival(true, resources, ships, [defenderMessage, ownerMessage], [])
  }

  /**
   * Process ACS defend mission return
   */
  async processReturn(context: MissionContext): Promise<MissionReturnResult> {
    const { mission, originPlanet } = context

    if (!originPlanet) {
      return this.errorReturn('Origin planet not found')
    }

    const ships = this.getShips(mission)
    const resources = this.getResources(mission)

    // Add ships back to origin
    const addShipsResult = await this.addShipsToPlanet(originPlanet.id, ships)
    if (!addShipsResult.success) {
      return this.errorReturn(`Failed to return ships: ${addShipsResult.error}`)
    }

    // Add resources
    if (hasResources(resources)) {
      const addResourcesResult = await this.addResourcesToPlanet(originPlanet.id, resources)
      if (!addResourcesResult.success) {
        return this.errorReturn(`Failed to return resources: ${addResourcesResult.error}`)
      }
    }

    // Update participant status if part of ACS operation
    await this.supabase
      .from('acs_participants')
      .update({ status: 'returned' })
      .eq('fleet_mission_id', mission.id)

    const message = this.createMessage(
      mission.user_id,
      'battle',
      'Defense Fleet Returned',
      `Your defense fleet has returned from ${this.formatCoords(
        mission.destination_galaxy,
        mission.destination_system,
        mission.destination_position
      )}.`
    )

    return this.successReturn([message], [])
  }
}

File: unity_client/Assets/Scripts/SharedLib/Missions/Handlers/ACSDefendMission.cs
// =============================================================================
// ACSDefendMission.cs — Allied Combat System defend mission handler
// Port of: src/lib/missions/handlers/ACSDefendMission.ts
// Namespace: OGameX.SharedLib.Missions.Handlers
// =============================================================================

using System;
using System.Collections.Generic;

namespace OGameX.SharedLib.Missions.Handlers
{
    /// <summary>
    /// Handles ACS (Alliance/Allied Combat System) defense missions.
    /// 
    /// Defense fleets join a target planet's forces temporarily:
    /// 1. Fleet arrives at the allied planet
    /// 2. Ships join the planetary defense for a hold period
    /// 3. If attacked during hold, ships participate in defense
    /// 4. After hold time expires, fleet returns automatically
    /// 
    /// ACS defend is different from deployment because:
    /// - Ships remain under original owner's control
    /// - Ships return after hold period (default: 12 hours)
    /// - Can be sent to ANY player's planet (with invitation)
    /// </summary>
    public class ACSDefendMission : BaseMission
    {
        public override MissionType HandlerMissionType => MissionType.ACSDefend;
        public override bool HasReturn => true;
        public override string Name => "ACS Defend";

        // =====================================================================
        // CONSTANTS
        // =====================================================================

        /// <summary>Default hold time at target (in hours).</summary>
        public const double DefaultHoldTimeHours = 12.0;

        /// <summary>Maximum hold time at target (in hours).</summary>
        public const double MaxHoldTimeHours = 32.0;

        // =====================================================================
        // ARRIVAL PROCESSING
        // =====================================================================

        public override MissionArrivalResult ProcessArrival(MissionContext context)
        {
            var mission = context.Mission;
            var targetPlanet = context.TargetPlanet;

            if (targetPlanet == null)
                return ErrorArrival("Target planet not found");

            var ships = GetShips(mission);
            var originCoords = FormatCoords(
                mission.OriginGalaxy, mission.OriginSystem, mission.OriginPosition);
            var targetCoords = FormatCoords(
                mission.DestinationGalaxy, mission.DestinationSystem, mission.DestinationPosition);

            // Calculate hold duration
            double holdHours = GetHoldDuration(context);
            holdHours = Math.Min(holdHours, MaxHoldTimeHours);

            // Build messages
            var messages = new List<MissionMessage>();

            // Notify fleet owner
            messages.Add(CreateMessage(
                mission.UserId,
                MessageType.System,
                "ACS Defense - Fleet Arrived",
                $"Your defense fleet has arrived at {targetCoords} and will hold " +
                $"position for {holdHours:F1} hours.\n\n" +
                $"Ships stationed:\n{ships.FormatList()}"
            ));

            // Notify planet owner (if different)
            if (targetPlanet.UserId != mission.UserId)
            {
                messages.Add(CreateMessage(
                    targetPlanet.UserId,
                    MessageType.System,
                    "ACS Defense - Fleet Arrived",
                    $"An allied fleet from {originCoords} has arrived to defend your planet at {targetCoords}.\n\n" +
                    $"Ships:\n{ships.FormatList()}\n" +
                    $"Hold duration: {holdHours:F1} hours"
                ));
            }

            // The fleet will return after hold time
            // The game layer schedules the return mission with the hold delay
            return SuccessArrival(
                shouldReturn: true,
                returnResources: GetResources(mission),
                returnShips: ships,
                messages: messages
            );
        }

        // =====================================================================
        // RETURN PROCESSING
        // =====================================================================

        public override MissionReturnResult ProcessReturn(MissionContext context)
        {
            var mission = context.Mission;
            var resources = GetResources(mission);

            var originCoords = FormatCoords(
                mission.DestinationGalaxy, mission.DestinationSystem, mission.DestinationPosition);
            var fromCoords = FormatCoords(
                mission.OriginGalaxy, mission.OriginSystem, mission.OriginPosition);

            var message = CreateMessage(
                mission.UserId,
                MessageType.System,
                "ACS Defense - Fleet Returned",
                $"Your defense fleet has completed its station at {fromCoords} " +
                $"and returned to {originCoords}."
            );

            return SuccessReturn(new List<MissionMessage> { message });
        }

        // =====================================================================
        // VIRTUAL HOOKS
        // =====================================================================

        /// <summary>
        /// Get the hold duration in hours for this ACS defense.
        /// Override to read from mission data or ACS group settings.
        /// </summary>
        protected virtual double GetHoldDuration(MissionContext context)
        {
            return DefaultHoldTimeHours;
        }
    }
}
